File: MoodDiary.java
// MoodDiary.java
import java.io.*;
import java.nio.file.*;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.stream.Collectors;

record MoodEntry(int id, String timestamp, int mood, String note) implements Serializable {}

class MoodDiary implements Serializable {
    private static final long serialVersionUID = 1L;
    private List<MoodEntry> entries = new ArrayList<>();
    private int nextId = 1;

    public MoodEntry addEntry(int mood, String note) {
        if (mood < 1 || mood > 10) {
            throw new IllegalArgumentException("Оценка должна быть от 1 до 10");
        }
        MoodEntry entry = new MoodEntry(
            nextId,
            LocalDateTime.now().format(DateTimeFormatter.ISO_LOCAL_DATE_TIME),
            mood,
            note
        );
        entries.add(entry);
        nextId++;
        return entry;
    }

    public Optional<MoodEntry> findEntry(int id) {
        return entries.stream().filter(e -> e.id() == id).findFirst();
    }

    public boolean deleteEntry(int id) {
        return entries.removeIf(e -> e.id() == id);
    }

    public Map<String, Object> getStats() {
        int total = entries.size();
        if (total == 0) {
            return Map.of("total", 0, "average", null, "min", null, "max", null);
        }
        IntSummaryStatistics stats = entries.stream().mapToInt(MoodEntry::mood).summaryStatistics();
        return Map.of(
            "total", total,
            "average", stats.getAverage(),
            "min", stats.getMin(),
            "max", stats.getMax()
        );
    }

    public void saveToFile(String filename) throws IOException {
        try (ObjectOutputStream oos = new ObjectOutputStream(Files.newOutputStream(Paths.get(filename)))) {
            oos.writeObject(this);
        }
    }

    public void loadFromFile(String filename) throws IOException, ClassNotFoundException {
        try (ObjectInputStream ois = new ObjectInputStream(Files.newInputStream(Paths.get(filename)))) {
            MoodDiary loaded = (MoodDiary) ois.readObject();
            this.entries = loaded.entries;
            this.nextId = loaded.nextId;
        }
    }

    public List<MoodEntry> getEntries() { return Collections.unmodifiableList(entries); }
}

public class MoodDiaryApp {
    private static final Scanner scanner = new Scanner(System.in);

    private static String readString(String prompt) {
        System.out.print(prompt);
        return scanner.nextLine().trim();
    }

    private static int readInt(String prompt) {
        while (true) {
            try {
                System.out.print(prompt);
                return Integer.parseInt(scanner.nextLine().trim());
            } catch (NumberFormatException e) {
                System.out.println("Введите число.");
            }
        }
    }

    private static void printEntry(MoodEntry entry) {
        String emoji = entry.mood() >= 7 ? "😄" : entry.mood() >= 4 ? "😐" : "😞";
        System.out.printf("%s #%d - %s - Оценка: %d/10%n", emoji, entry.id(), entry.timestamp(), entry.mood());
        System.out.printf("   Заметка: %s%n", entry.note());
    }

    public static void main(String[] args) {
        MoodDiary diary = new MoodDiary();
        try {
            diary.loadFromFile("mood_data.ser");
        } catch (IOException | ClassNotFoundException e) {
            System.out.println("Не удалось загрузить данные, начинаем с пустого дневника.");
        }

        while (true) {
            System.out.println("\n===== ДНЕВНИК НАСТРОЕНИЯ (Java) =====");
            System.out.println("1. Добавить запись");
            System.out.println("2. Показать все записи");
            System.out.println("3. Показать статистику");
            System.out.println("4. Удалить запись");
            System.out.println("5. Сохранить в файл");
            System.out.println("6. Загрузить из файла");
            System.out.println("0. Выход");
            String choice = readString("Выберите действие: ");

            switch (choice) {
                case "0" -> { return; }
                case "1" -> {
                    int mood = readInt("Оценка настроения (от 1 до 10): ");
                    if (mood < 1 || mood > 10) {
                        System.out.println("Оценка должна быть от 1 до 10.");
                        continue;
                    }
                    String note = readString("Заметка: ");
                    MoodEntry entry = diary.addEntry(mood, note);
                    System.out.printf("Запись добавлена с ID %d%n", entry.id());
                }
                case "2" -> {
                    if (diary.getEntries().isEmpty()) {
                        System.out.println("Нет записей.");
                    } else {
                        diary.getEntries().forEach(MoodDiaryApp::printEntry);
                    }
                }
                case "3" -> {
                    var stats = diary.getStats();
                    if ((int)stats.get("total") == 0) {
                        System.out.println("Нет записей для статистики.");
                    } else {
                        System.out.println("\n=== СТАТИСТИКА ===");
                        System.out.printf("Всего записей: %d%n", stats.get("total"));
                        System.out.printf("Средняя оценка: %.2f%n", stats.get("average"));
                        System.out.printf("Минимальная оценка: %d%n", stats.get("min"));
                        System.out.printf("Максимальная оценка: %d%n", stats.get("max"));
                    }
                }
                case "4" -> {
                    int id = readInt("Введите ID записи для удаления: ");
                    if (diary.deleteEntry(id)) {
                        System.out.println("Запись удалена.");
                    } else {
                        System.out.println("Запись не найдена.");
                    }
                }
                case "5" -> {
                    try {
                        diary.saveToFile("mood_data.ser");
                        System.out.println("Сохранено.");
                    } catch (IOException e) {
                        System.out.println("Ошибка сохранения: " + e.getMessage());
                    }
                }
                case "6" -> {
                    try {
                        diary.loadFromFile("mood_data.ser");
                        System.out.println("Загружено.");
                    } catch (IOException | ClassNotFoundException e) {
                        System.out.println("Ошибка загрузки: " + e.getMessage());
                    }
                }
                default -> System.out.println("Неизвестная команда.");
            }
        }
    }
}

File: MoodDiary.cs
// MoodDiary.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

public record MoodEntry(
    int Id,
    string Timestamp,
    int Mood,
    string Note
);

public class MoodDiary
{
    private List<MoodEntry> entries = new();
    private int nextId = 1;

    public IReadOnlyList<MoodEntry> Entries => entries.AsReadOnly();

    public MoodEntry AddEntry(int mood, string note)
    {
        if (mood < 1 || mood > 10)
            throw new ArgumentException("Оценка должна быть от 1 до 10");
        var entry = new MoodEntry(
            nextId,
            DateTime.Now.ToString("o"),
            mood,
            note
        );
        entries.Add(entry);
        nextId++;
        return entry;
    }

    public MoodEntry? FindEntry(int id) => entries.FirstOrDefault(e => e.Id == id);

    public bool DeleteEntry(int id) => entries.RemoveAll(e => e.Id == id) > 0;

    public Dictionary<string, object> GetStats()
    {
        int total = entries.Count;
        if (total == 0)
            return new Dictionary<string, object> { ["total"] = 0, ["average"] = null, ["min"] = null, ["max"] = null };
        var moods = entries.Select(e => e.Mood);
        return new Dictionary<string, object>
        {
            ["total"] = total,
            ["average"] = moods.Average(),
            ["min"] = moods.Min(),
            ["max"] = moods.Max()
        };
    }

    public void SaveToFile(string filename)
    {
        var options = new JsonSerializerOptions { WriteIndented = true };
        string json = JsonSerializer.Serialize(entries, options);
        File.WriteAllText(filename, json);
    }

    public void LoadFromFile(string filename)
    {
        if (!File.Exists(filename)) return;
        string json = File.ReadAllText(filename);
        var loaded = JsonSerializer.Deserialize<List<MoodEntry>>(json);
        if (loaded != null)
        {
            entries = loaded;
            nextId = entries.Any() ? entries.Max(e => e.Id) + 1 : 1;
        }
    }
}

public static class Program
{
    private static string ReadString(string prompt)
    {
        Console.Write(prompt);
        return Console.ReadLine()?.Trim() ?? "";
    }

    private static int ReadInt(string prompt)
    {
        while (true)
        {
            Console.Write(prompt);
            if (int.TryParse(Console.ReadLine(), out int result))
                return result;
            Console.WriteLine("Введите число.");
        }
    }

    private static void PrintEntry(MoodEntry entry)
    {
        string emoji = entry.Mood >= 7 ? "😄" : entry.Mood >= 4 ? "😐" : "😞";
        Console.WriteLine($"{emoji} #{entry.Id} - {entry.Timestamp} - Оценка: {entry.Mood}/10");
        Console.WriteLine($"   Заметка: {entry.Note}");
    }

    public static void Main()
    {
        var diary = new MoodDiary();
        try { diary.LoadFromFile("mood_data.json"); }
        catch { Console.WriteLine("Не удалось загрузить данные."); }

        while (true)
        {
            Console.WriteLine("\n===== ДНЕВНИК НАСТРОЕНИЯ (C#) =====");
            Console.WriteLine("1. Добавить запись");
            Console.WriteLine("2. Показать все записи");
            Console.WriteLine("3. Показать статистику");
            Console.WriteLine("4. Удалить запись");
            Console.WriteLine("5. Сохранить в файл");
            Console.WriteLine("6. Загрузить из файла");
            Console.WriteLine("0. Выход");
            string choice = ReadString("Выберите действие: ");

            switch (choice)
            {
                case "0": return;
                case "1":
                    int mood = ReadInt("Оценка настроения (от 1 до 10): ");
                    if (mood < 1 || mood > 10)
                    {
                        Console.WriteLine("Оценка должна быть от 1 до 10.");
                        continue;
                    }
                    string note = ReadString("Заметка: ");
                    var entry = diary.AddEntry(mood, note);
                    Console.WriteLine($"Запись добавлена с ID {entry.Id}");
                    break;
                case "2":
                    if (!diary.Entries.Any()) Console.WriteLine("Нет записей.");
                    else foreach (var e in diary.Entries) PrintEntry(e);
                    break;
                case "3":
                    var stats = diary.GetStats();
                    if ((int)stats["total"] == 0)
                        Console.WriteLine("Нет записей для статистики.");
                    else
                    {
                        Console.WriteLine("\n=== СТАТИСТИКА ===");
                        Console.WriteLine($"Всего записей: {stats["total"]}");
                        Console.WriteLine($"Средняя оценка: {stats["average"]:F2}");
                        Console.WriteLine($"Минимальная оценка: {stats["min"]}");
                        Console.WriteLine($"Максимальная оценка: {stats["max"]}");
                    }
                    break;
                case "4":
                    int id = ReadInt("Введите ID записи для удаления: ");
                    if (diary.DeleteEntry(id)) Console.WriteLine("Запись удалена.");
                    else Console.WriteLine("Запись не найдена.");
                    break;
                case "5":
                    try { diary.SaveToFile("mood_data.json"); Console.WriteLine("Сохранено."); }
                    catch (Exception ex) { Console.WriteLine($"Ошибка: {ex.Message}"); }
                    break;
                case "6":
                    try { diary.LoadFromFile("mood_data.json"); Console.WriteLine("Загружено."); }
                    catch (Exception ex) { Console.WriteLine($"Ошибка: {ex.Message}"); }
                    break;
                default: Console.WriteLine("Неизвестная команда."); break;
            }
        }
    }
}
